packages/tools/devtools/devtools-core/src/data-visualization/SharedTreeVisualizer.ts

```diff
@@ -148,64 +155,113 @@ export function toVisualTree(tree: VisualSharedTreeNode): VisualChildNode {
 }
 
 /**
- * Concatenrate allowed types for `ObjectNodeStoredSchema` and `MapNodeStoredSchema`.
+ * Concatenate allowed types for `ObjectNodeStoredSchema` and `MapNodeStoredSchema`.
  */
-function concatenateTypes(fieldTypes: ReadonlySet<string>): string {
+export function concatenateTypes(fieldTypes: ReadonlySet<string>): string {
 	return [...fieldTypes].join(" | ");
 }
 
 /**
- * Returns the allowed fields & types for the object fields (e.g., `foo : string | number, bar: boolean`)
+ * Properties that describe schema constraints for a field in the tree
  */
-function getObjectAllowedTypes(schema: SimpleObjectNodeSchema): string {
-	const result: string[] = [];
+interface FieldSchemaProperties {
+	/**
+	 * Set of type names that are valid for this specific node position in the tree.
+	 * This is a subset of the types defined in treeDefinitions.
+	 */
+	allowedTypes?: ReadonlySet<string>;
```

Comment:
Do the properties here need to be optional?

Reply:
I want it to align the optionality with that of `VisualSharedTreeTypes.ts`.

Reply:
I think (Alejandro please correct me if I'm wrong), the suggestion is that we could make them required but allow undefined. Since these are private implementation details, we don't need to worry as much about end-user ergonomics. And making the properties required ensures we don't forget to set it when we want to.

Reply:
We should also probably make these properties `readonly` while we're here.

Reply:
Yeah, for non-exported interfaces (not part of public API), I think in most cases we benefit from making things explicit.

Reply:
Removed optional flag to both `allowedTypes` and `required`.